autoconfigure/storage-elasticsearch-aws/src/main/java/zipkin/autoconfigure/storage/elasticsearch/aws/AWSSignatureVersion4.java

```diff
@@ -78,68 +75,50 @@
-          if (hosts.size() > 1) throw new RuntimeException("Expected one AWS ES host: " + hosts);
-          host = URI.create(hosts.get(0)).getHost();
-        }
-      }
-    }
-    return host;
-  }
-
   @Override public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
-    return HttpResponse.from(
-        req.aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc())
-            .thenCompose(aggReg -> {
-              try {
-                AggregatedHttpRequest signed = sign(ctx, aggReg);
-                return delegate().execute(ctx, HttpRequest.of(signed)).aggregate();
-              } catch (Exception e) {
-                CompletableFuture<AggregatedHttpResponse> future = new CompletableFuture<>();
-                future.completeExceptionally(e);
-                return future;
-              }
-            })
-            .thenApply(aggResp -> {
-              if (!aggResp.status().equals(HttpStatus.FORBIDDEN)) {
-                return HttpResponse.of(aggResp);
-              }
-              String body = aggResp.contentUtf8();
-              try {
-                JsonParser message = enterPath(jsonParser(body), "message");
-                if (message != null) throw new IllegalStateException(message.getValueAsString());
-              } catch (IOException e) {
-                // Ignore JSON parse failure.
-              }
-              throw new IllegalStateException(aggResp.contentUtf8());
-            }));
+    // We aggregate the reqiest body with pooled objects because signing implies reading it before
+    // sending it to Elasticsearch.
+    return HttpResponse.from(req.aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc())
+        .thenCompose(aggReg -> {
+          try {
+            AggregatedHttpRequest signed = sign(ctx, aggReg);
+            return delegate().execute(ctx, HttpRequest.of(signed))
+                // We aggregate the response with pooled objects because it could be large. This
+                // reduces heap usage when parsing json or when http body logging is enabled.
+                .aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc());
+          } catch (Exception e) {
+            CompletableFuture<AggregatedHttpResponse> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+          }
+        })
+        .thenApply(aggResp -> {
+          if (!aggResp.status().equals(HttpStatus.FORBIDDEN)) return HttpResponse.of(aggResp);
+
+          // We only set a body-related message when it Amazon's format
+          StringBuilder message = new StringBuilder().append(req.path()).append(" failed: ");
+          String awsMessage = null;
+          try (InputStream stream = aggResp.content().toInputStream()) {
+            awsMessage = JSON.readTree(stream).at("/message").textValue();
+          } catch (IOException e) {
+            // Ignore JSON parse failure.
+          }
```

Comment:
```java
finally {
  ReferenceCountUtil.safeRelease(aggResp.content())
}
```

In the end, we decided that `toInputStream` creates an additional reference instead of itself releasing `content()` completely.